Increase precision of CIEXYZ conversion matrix

m13253 opened the pull request:
## Summary of the Pull Request

**What is this about:** CIEXYZ output has 4 decimal places, so the conversion matrix should be more than 6 digits to avoid round-off errors.

**What is included in the PR:** The sRGB to CIEXYZ-D65 conversion matrix has been updated with a higher precision one.

**How does someone test / validate:** Issue #17040 contains testing data.

## Quality Checklist

- [ ] **Linked issue:** #17040
- [ ] **Communication:** I've discussed this with core contributors in the issue. 
- [ ] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [ ] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

m13253 committed:
Increase precision of CIEXYZ conversion matrix

The output has 4 decimal places, so the conversion matrix should be more than 6 digits to avoid round-off errors.

Fix #17040

m13253 commented:
**Disclaimer:**
I do not have a C# development environment on my computer, therefore, I have not tested the code.
However, I believe the updated code should theoretically work, since the only change is the numbers.

m13253 commented:
> I think you should update the tests too.

I plan to spend some time to generate the test results using other color management systems (e.g. LittleCMS2).
There is no meaning to test a algorithm with its own output, if the algorithm itself is inaccurate.

I already confirmed, using my new matrix, at least 6 colors are accurate. But there are 30 colors.
I will need to write some script to do this automatically. So I will come back some days later.

htcfreek commented:
> > I think you should update the tests too.
> 
> I plan to spend some time to generate the test results using other color management systems (e.g. LittleCMS2).
> There is no meaning to test a algorithm with its own output, if the algorithm itself is inaccurate.
> 
> I already confirmed, using my new matrix, at least 6 colors are accurate. But there are 30 colors.
> I will need to write some script to do this automatically. So I will come back some days later.

What I mean is that we have test to make sure that the code gaves the expected output. And because you changed the calculation they are falling now

m13253 commented:
Yes, I understand.

What I mean is, I think I need to calculate the same result using at least one 3rd-party software, then type the expected results into the source code, to make sure my calculations are correct.

The previous test is more like a regression test (instead of a correctness test), as it compares the algorithm with itself, and cannot verify the correctness of the algorithm.

I am currently doing this verification step, and I will push another commit today or tomorrow. I didn't realize there is such a test in the source code. So please allow me some time to make sure when this PR gets merged, all results are correct.

By the way, since the CIELAB conversion depends on CIEXYZ, I have to change and verify both tests.

m13253 committed:
Match unit tests and docs with new CIEXYZ conversion matrix

m13253 committed:
Remove negative sign from zeros

I generated the unit test results from other color-management systems.
It seems that they sometimes output negative zeros for very small values.
Let's just remove the negative signs for aesthetic.

m13253 committed:
Fix spelling mistakes in ColorConverterTest.cs

m13253 committed:
Explain how to obtain CIEXYZ unit test reference values

m13253 committed:
Explain the CIELAB output is D65 adapted version

jaimecbernardo commented:
@RubenFricke , This change seems like it'll improve the results.
Any comments here, since you were the one to bring this color space in https://github.com/microsoft/PowerToys/pull/12935 ?

RubenFricke commented:
I checked the code and the new values. Everything looks good, a nice improvement in accuracy. Well done!

Online I found some other people who calculated this value with an extra decimal (e.g. 01933081871559185 -> 019330818715591851 & 07219231536073372 -> 072192315360733715), although I'm fairly sure that you wouldn't see any differences in the output :smile: . @jaimecbernardo , are there any guidelines for the accuracy of constants in calculations? Or does the developer decide this depending on the use case? 

Thanks for improving this calculation @m13253!

m13253 commented:
> Online I found some other people who calculated this value with an extra decimal

I calculated the values using 128-bit float-point precision. [(Algorithm)](https://github.com/m13253/colorspace-routines)
That's 19 more decimal digits, so I'm sure my numbers are better than theirs. (smile)

The current CIEXYZ output in PowerToys has 4 decimal digits, and CIELAB 2 digits.
Different other software has different output precision.
For example, GIMP has 1 digit LAB, Affinity has 0 digits LAB, Krita has 2 digits XYZ/LAB.
Therefore, I believe the new output is precise enough now.

Furthermore, the new code now makes sure that neutral grays have exactly a=0.00, b=0.00.

By the way, how do I deal with the spell-checker that blocks the merge?
Should I just ignore it or do something to get rid of it?

m13253 commented:
> I guess there will always be off by 1 errors here.

I confirm even if I add one more digit, the binary representation will be the same given `double` as the data type.

17 decimal digits are what `double` can offer.

m13253 commented:
> This seems to be caused by this line: `http://www.brucelindbloom.com/index.html?Eqn_RGB_XYZ_Matrix.html`
> 
> 
> 
> For now, I'd say to add `brucelindbloom` and `Eqn` to `.github\actions\spell-check\expect.txt`.

Also `transicc`, which is the name of a 3rd-party program I mentioned in the comment. I wrote that comment in the unit test file.

I also checked the bypass list. The bypass list only bypasses **some** URLs, like Wikipedia and MSDN. Is this the expected behavior to not bypass **all** URLs?

jaimecbernardo commented:
Perhaps add a bypass to `http://www.brucelindbloom.com`, then? That sounds the more accurate route here. Thank you!

jsoref commented:
I'd just as the entry to expect. It really only makes sense to use a pattern if there are a bunch of items or if they're really garbage (e.g. hex runs).

m13253 committed:
Add words related to CIEXYZ conversion to spellcheck bypass list

m13253 commented:
> Perhaps add a bypass to `http://www.brucelindbloom.com`, then? That sounds the more accurate route here. Thank you!

I updated `expect.txt` to include `brucelindbloom`, `Eqn`, `transicc`.
Please check the changeset to see if it can be approved.

jaimecbernardo reviewed (approved):
LGTM! Thank you for the contribution!
